### PR TITLE
Update Binaryen

### PR DESCRIPTION
**Pull request by dcodeIO**
As usual, this gets us up to date with the latest changes in Binaryen. Quite a bit of chaos here with missing C-APIs that may be good to tackle.

- [x] I've read the contributing guidelines

### CODE REVIEW

**Comment by dcodeIO** [comment]
Some notes I made so I don't forget

```
SIMDWidenOp enum
PrefetchOp enum
RefAsOp enum
BrOnOp enum
_BinaryenRefIsGetOp
_BinaryenRefIsSetOp

BinaryenTableGetInitial -> int is Index
BinaryenTableGetMax -> int is Index

BrOnExn removed, BrOn missing now

SetFunctionTable before AddTable/ImportTable breaks
```

plus all the `TODO_Binaryen` constants.

**Comment by MaxGraey** [comment]
Seems it still slower on 1.7 times for binaryen's optimization. Could we create issue about this in Binaryen repo?

**Comment by dcodeIO** [comment]
I haven't investigated the slowdown myself, so I guess you have more information on it than I have. Could you make one? :)

**Comment by AurelienRibon** [comment]
Impressive work. How to do implement such migration? Updating the lib version and seeing what fails?

**Comment by dcodeIO** [comment]
My process there is typically:
* Blindly upgrade the Binaryen dependency to the latest version
* Go to `src/binaryen-c.h` in the Binaryen repo and click "History"
* Work my way through what has changed or is new
  * Update src/glue/binaryen.d.ts
  * Refactor name and API changes in the compiler to mirror the d.ts
* Run `scripts/update-constants` to sync the constants, which will print errors if something is still off
* Run the test suite and evaluate fixture changes
* Bonus: Implement new features

**Comment by AurelienRibon** [comment]
Interesting, thank you, that's really a nice update process.

**Comment by MaxGraey** [comment]
I'm wondering may be this https://github.com/WebAssembly/binaryen/pull/3609 fix issue with not always de-dupllicate functions?

**Comment by dcodeIO** [comment]
Going to check once we have a build :)

File: src/flow.ts [hunk 1]
```diff
@@ -390,18 +392,30 @@ export class Flow {
...
+        else parentFunction.tempEqrefs = temps = [];
+        break;
+      }
+      case <u32>NativeType.I31ref: {
+        let tempI31refs = parentFunction.tempI31refs;
+        if (tempI31refs) temps = tempI31refs;
+        else parentFunction.tempI31refs = temps = [];
+        break;
+      }
+      case <u32>NativeType.Dataref: {
```
Comment (MaxGraey) [inline src/flow.ts:413]: btw I'm wondering why we need case to `u32` for switch all cases?

Comment (dcodeIO) [reply src/flow.ts:413]: Yes, because `NativeType` is `usize` (can be a pointer to a tuple for example), but all the canonical ones fit into 32 bits.

**Comment by tlively** [comment]
@MaxGraey it would be great if you could file a Binaryen issue for the slowdown.

**Comment by dcodeIO** [comment]
> I'm wondering may be this [WebAssembly/binaryen#3609](https://github.com/WebAssembly/binaryen/pull/3609) fix issue with not always de-dupllicate functions?

Seems the fixtures didn't change

**Comment by dcodeIO** [comment]
Making myself a handy list of SIMD instructions so I don't forget to check one

// Memory

- [ ] v128.load
- [ ] v128.load8x8_s
- [ ] v128.load8x8_u
- [ ] v128.load16x4_s
- [ ] v128.load16x4_u
- [ ] v128.load32x2_s
- [ ] v128.load32x2_u
- [ ] v128.load8_splat
- [ ] v128.load16_splat
- [ ] v128.load32_splat
- [ ] v128.load64_splat
- [ ] v128.store

// Basic

- [ ] v128.const
- [ ] i8x16.shuffle
- [ ] i8x16.swizzle

// Splat

- [ ] i8x16.splat
- [ ] i16x8.splat
- [ ] i32x4.splat
- [ ] i64x2.splat
- [ ] f32x4.splat
- [ ] f64x2.splat

// Lane

- [ ] i8x16.extract_lane_s
- [ ] i8x16.extract_lane_u
- [ ] i8x16.replace_lane
- [ ] i16x8.extract_lane_s
- [ ] i16x8.extract_lane_u
- [ ] i16x8.replace_lane
- [ ] i32x4.extract_lane
- [ ] i32x4.replace_lane
- [ ] i64x2.extract_lane
- [ ] i64x2.replace_lane
- [ ] f32x4.extract_lane
- [ ] f32x4.replace_lane
- [ ] f64x2.extract_lane
- [ ] f64x2.replace_lane

// Cmp

- [ ] i8x16.eq
- [ ] i16x8.eq
- [ ] i32x4.eq
- [ ] i8x16.ne
- [ ] i16x8.ne
- [ ] i32x4.ne
- [ ] i8x16.lt_s
- [ ] i16x8.lt_s
- [ ] i32x4.lt_s
- [ ] i8x16.lt_u
- [ ] i16x8.lt_u
- [ ] i32x4.lt_u
- [ ] i8x16.gt_s
- [ ] i16x8.gt_s
- [ ] i32x4.gt_s
- [ ] i8x16.gt_u
- [ ] i16x8.gt_u
- [ ] i32x4.gt_u
- [ ] i8x16.le_s
- [ ] i16x8.le_s
- [ ] i32x4.le_s
- [ ] i8x16.le_u
- [ ] i16x8.le_u
- [ ] i32x4.le_u
- [ ] i8x16.ge_s
- [ ] i16x8.ge_s
- [ ] i32x4.ge_s
- [ ] i8x16.ge_u
- [ ] i16x8.ge_u
- [ ] i32x4.ge_u
- [ ] f32x4.eq
- [ ] f64x2.eq
- [ ] f32x4.ne
- [ ] f64x2.ne
- [ ] f32x4.lt
- [ ] f64x2.lt
- [ ] f32x4.gt
- [ ] f64x2.gt
- [ ] f32x4.le
- [ ] f64x2.le
- [ ] f32x4.ge
- [ ] f64x2.ge

// v128

- [ ] v128.not
- [ ] v128.and
- [ ] v128.andnot
- [ ] v128.or
- [ ] v128.xor
- [ ] v128.bitselect
- [ ] v128.any_true

// Load lane

- [ ] v128.load8_lane
- [ ] v128.load16_lane
- [ ] v128.load32_lane
- [ ] v128.load64_lane
- [ ] v128.store8_lane
- [ ] v128.store16_lane
- [ ] v128.store32_lane
- [ ] v128.store64_lane
- [ ] v128.load32_zero
- [ ] v128.load64_zero

// Float conversion

- [ ] f32x4.demote_f64x2_zero
- [ ] f64x2.promote_low_f32x4

// iAxB

- [ ] i8x16.abs
- [ ] i16x8.abs
- [ ] i32x4.abs
- [ ] i64x2.abs
- [ ] i8x16.neg
- [ ] i16x8.neg
- [ ] i32x4.neg
- [ ] i64x2.neg
- [ ] i8x16.popcnt
- [ ] i16x8.q15mulr_sat_s
- [ ] i8x16.all_true
- [ ] i16x8.all_true
- [ ] i32x4.all_true
- [ ] i64x2.all_true
- [ ] i8x16.bitmask
- [ ] i16x8.bitmask
- [ ] i32x4.bitmask
- [ ] i64x2.bitmask
- [ ] i8x16.narrow_i16x8_s
- [ ] i16x8.narrow_i32x4_s
- [ ] i8x16.narrow_i16x8_u
- [ ] i16x8.narrow_i32x4_u
- [ ] f32x4.ceil
- [ ] i16x8.extend_low_i8x16_s
- [ ] i32x4.extend_low_i16x8_s
- [ ] i64x2.extend_low_i32x4_s
- [ ] f32x4.floor
- [ ] i16x8.extend_high_i8x16_s
- [ ] i32x4.extend_high_i16x8_s
- [ ] i64x2.extend_high_i32x4_s
- [ ] f32x4.trunc
- [ ] i16x8.extend_low_i8x16_u
- [ ] i32x4.extend_low_i16x8_u
- [ ] i64x2.extend_low_i32x4_u
- [ ] f32x4.nearest
- [ ] i16x8.extend_high_i8x16_u
- [ ] i32x4.extend_high_i16x8_u
- [ ] i64x2.extend_high_i32x4_u
- [ ] i8x16.shl
- [ ] i16x8.shl
- [ ] i32x4.shl
- [ ] i64x2.shl
- [ ] i8x16.shr_s
- [ ] i16x8.shr_s
- [ ] i32x4.shr_s
- [ ] i64x2.shr_s
- [ ] i8x16.shr_u
- [ ] i16x8.shr_u
- [ ] i32x4.shr_u
- [ ] i64x2.shr_u
- [ ] i8x16.add
- [ ] i16x8.add
- [ ] i32x4.add
- [ ] i64x2.add
- [ ] i8x16.add_sat_s
- [ ] i16x8.add_sat_s
- [ ] i8x16.add_sat_u
- [ ] i16x8.add_sat_u
- [ ] i8x16.sub
- [ ] i16x8.sub
- [ ] i32x4.sub
- [ ] i64x2.sub
- [ ] i8x16.sub_sat_s
- [ ] i16x8.sub_sat_s
- [ ] i8x16.sub_sat_u
- [ ] i16x8.sub_sat_u
- [ ] f64x2.ceil
- [ ] f64x2.nearest
- [ ] f64x2.floor
- [ ] i16x8.mul
- [ ] i32x4.mul
- [ ] i64x2.mul
- [ ] i8x16.min_s
- [ ] i16x8.min_s
- [ ] i32x4.min_s
- [ ] i64x2.eq
- [ ] i8x16.min_u
- [ ] i16x8.min_u
- [ ] i32x4.min_u
- [ ] i64x2.ne
- [ ] i8x16.max_s
- [ ] i16x8.max_s
- [ ] i32x4.max_s
- [ ] i64x2.lt_s
- [ ] i8x16.max_u
- [ ] i16x8.max_u
- [ ] i32x4.max_u
- [ ] i64x2.gt_s
- [ ] f64x2.trunc
- [ ] i32x4.dot_i16x8_s
- [ ] i64x2.le_s
- [ ] i8x16.avgr_u
- [ ] i16x8.avgr_u
- [ ] i64x2.ge_s
- [ ] i16x8.extadd_pairwise_i8x16_s
- [ ] i16x8.extmul_low_i8x16_s
- [ ] i32x4.extmul_low_i16x8_s
- [ ] i64x2.extmul_low_i32x4_s
- [ ] i16x8.extadd_pairwise_i8x16_u
- [ ] i16x8.extmul_high_i8x16_s
- [ ] i32x4.extmul_high_i16x8_s
- [ ] i64x2.extmul_high_i32x4_s
- [ ] i32x4.extadd_pairwise_i16x8_s
- [ ] i16x8.extmul_low_i8x16_u
- [ ] i32x4.extmul_low_i16x8_u
- [ ] i64x2.extmul_low_i32x4_u
- [ ] i32x4.extadd_pairwise_i16x8_u
- [ ] i16x8.extmul_high_i8x16_u
- [ ] i32x4.extmul_high_i16x8_u
- [ ] i64x2.extmul_high_i32x4_u

// fAxB

- [ ] f32x4.abs
- [ ] f64x2.abs
- [ ] f32x4.neg
- [ ] f64x2.neg
- [ ] f32x4.sqrt
- [ ] f64x2.sqrt
- [ ] f32x4.add
- [ ] f64x2.add
- [ ] f32x4.sub
- [ ] f64x2.sub
- [ ] f32x4.mul
- [ ] f64x2.mul
- [ ] f32x4.div
- [ ] f64x2.div
- [ ] f32x4.min
- [ ] f64x2.min
- [ ] f32x4.max
- [ ] f64x2.max
- [ ] f32x4.pmin
- [ ] f64x2.pmin
- [ ] f32x4.pmax
- [ ] f64x2.pmax

// Conversion

- [ ] i32x4.trunc_sat_f32x4_s
- [ ] i32x4.trunc_sat_f32x4_u
- [ ] f32x4.convert_i32x4_s
- [ ] f32x4.convert_i32x4_u
- [ ] i32x4.trunc_sat_f64x2_s_zero
- [ ] i32x4.trunc_sat_f64x2_u_zero
- [ ] f64x2.convert_low_i32x4_s
- [ ] f64x2.convert_low_i32x4_u

File: std/assembly/reference.ts [hunk 2]
```diff
@@ -20,6 +20,12 @@ export abstract class Eqref extends Ref {
 
 @final @unmanaged
 export abstract class I31ref extends Ref {
+  get(this: i31ref): i32 {
+    return i31.get(this);
+  }
+  getUnsigned(this: i31ref): u32 {
+    return <u32>i31.get(this);
+  }
```
Comment (MaxGraey) [inline std/assembly/reference.ts:28]: I think it's unnecessary. It's much shorted use `<u32>ref.get()` instead `ref.getUnsigned()` I guess

Comment (dcodeIO) [reply std/assembly/reference.ts:28]: The problem there is that an `i31` does not properly sign-extend, so it is important how it is loaded. Say we are loading `0x7fffffff`, then loading that as signed becomes `-1`, but `<u32>-1` gives us `0xffffffff`, not `0x7fffffff`, as would have been the value when loading the `i31` as unsigned.

Comment (MaxGraey) [reply std/assembly/reference.ts:28]: I see. Hmm

What about:
```ts
class I31ref {
  signed(this: i31ref): i32 {
    return i31.get<i32>(this);
  }
  
  unsigned(this: i31ref): u32 {
    return i31.get<u32>(this);
  }
}
```
or
```ts
class I31ref {
  i32(this: i31ref): i32 {
    return i31.get<i32>(this);
  }
  
  u32(this: i31ref): u32 {
    return i31.get<u32>(this);
  }
}
```

Comment (dcodeIO) [reply std/assembly/reference.ts:28]: Wondering, what if we do this as casts instead:

```ts
var ref = i31.new(42);
var i32 = <i32>ref;
var u32 = <u32>ref;
```

Comment (MaxGraey) [reply std/assembly/reference.ts:28]: yeah. that's will be the best way

**Comment by dcodeIO** [comment]
I guess this one is fine to go now, given that the remaining constants and bindings are missing, we don't have a cast story for GC types yet, so I'll have to look into this again anyhow. The lists above should be useful then.

**Comment by MaxGraey** [comment]
@dcodeIO Thanks for this excellent job! I'll check little bit later. But I'm would like report about slowdown issue of binaryen. Try to bisect version when this happened

**Review by MaxGraey — APPROVED** [review]
I saw couple missed pieces for ref_is / ref_as. I guess this still not implemented on binaryen?

LGTM.

**Comment by dcodeIO** [comment]
Right, we cannot yet make proper use of these without an initial implementation of Wasm GC, but it's good to keep the bindings up to date already so everything's there once we can :)

**Pull request merged**